File: lib/indicators.js
'use strict';

// Rise / fall / warning are decided while a value is formatted, but drawn much
// later — either as an icon in the menu bar image or as a text glyph. So the
// formatter leaves a marker behind and this module turns it into whichever the
// user has chosen. Control characters are used as the markers because they
// cannot show up in a real response or in a user's template.

const MARKS = { rise: '\u0001', fall: '\u0002', warn: '\u0003' };
const MARK_PATTERN = /[\u0001-\u0003]/g;

const TEXT_GLYPH = {
  [MARKS.rise]: '▴',
  [MARKS.fall]: '▾',
  [MARKS.warn]: '⚠',
};

// Anywhere a value is shown as plain text — menu items, tooltips, the log, the
// clipboard — the markers become ordinary characters.
const toText = (value) =>
  String(value).replace(MARK_PATTERN, (mark) => TEXT_GLYPH[mark] || '');

const INDICATOR_STYLES = [
  { id: 'chevron', label: 'Chevron' },
  { id: 'arrow', label: 'Arrow' },
  { id: 'triangle', label: 'Triangle' },
  { id: 'text', label: 'Text' },
];

const DEFAULT_INDICATOR = 'chevron';

const isIndicatorStyle = (id) =>
  INDICATOR_STYLES.some((style) => style.id === id);

const normalizeIndicator = (id) =>
  isIndicatorStyle(id) ? id : DEFAULT_INDICATOR;

// Every icon is drawn on the same 10x10 grid so the styles line up with one
// another and with the text beside them.
const PATHS = {
  chevron: {
    size: 9,
    stroke: 1.7,
    up: 'M1.6 6.4 L5 3 L8.4 6.4',
    down: 'M1.6 3.6 L5 7 L8.4 3.6',
  },
  arrow: {
    size: 9,
    stroke: 1.6,
    up: 'M5 8.2 L5 2.2 M2.2 5 L5 2.2 L7.8 5',
    down: 'M5 1.8 L5 7.8 M2.2 5 L5 7.8 L7.8 5',
  },
  triangle: {
    size: 8,
    fill: true,
    stroke: 1.2,
    up: 'M5 2.3 L8.1 7.5 L1.9 7.5 Z',
    down: 'M5 7.7 L1.9 2.5 L8.1 2.5 Z',
  },
};

// A warning sign at the same weight as the direction icons.
const WARNING = {
  size: 10,
  stroke: 1.3,
  path: 'M5 1.7 L9.1 8.6 L0.9 8.6 Z M5 4.2 L5 6.1 M5 7.4 L5 7.5',
};

const svg = (px, stroke, d, fill) =>
  `<svg width="${px}" height="${px}" viewBox="0 0 10 10">` +
  `<path fill="${fill ? 'currentColor' : 'none'}" stroke="currentColor"` +
  ` stroke-width="${stroke}" stroke-linecap="round" stroke-linejoin="round"` +
  ` d="${d}"/></svg>`;

const iconSvg = (styleId, mark, scale) => {
  if (mark === MARKS.warn) {
    return svg(Math.round(WARNING.size * scale), WARNING.stroke, WARNING.path);
  }
  const spec = PATHS[normalizeIndicator(styleId)] || PATHS.chevron;
  return svg(
    Math.round(spec.size * scale),
    spec.stroke,
    mark === MARKS.fall ? spec.down : spec.up,
    spec.fill
  );
};

module.exports = {
  DEFAULT_INDICATOR,
  INDICATOR_STYLES,
  MARKS,
  MARK_PATTERN,
  iconSvg,
  isIndicatorStyle,
  normalizeIndicator,
  toText,
};

File: src-tauri/src/engine/indicators.rs
// Rise, fall and warning are decided while a value is formatted but shown much
// later, so the formatter leaves a control character behind and this module
// turns it into a glyph wherever text is drawn — menu items, tooltips, the
// tray title, the log and the clipboard. Control characters are used because
// they cannot appear in a real response or in a user's template.

pub const MARK_RISE: char = '\u{0001}';
pub const MARK_FALL: char = '\u{0002}';
pub const MARK_WARN: char = '\u{0003}';

pub fn is_mark(c: char) -> bool {
    c == MARK_RISE || c == MARK_FALL || c == MARK_WARN
}

/// The styles offered in the tray menu, in menu order. 1.x drew these as
/// icons; here each is a pair of characters the system font already has, so
/// the choice still shows up wherever a value does.
pub const STYLES: [(&str, &str); 4] = [
    ("chevron", "Chevron"),
    ("arrow", "Arrow"),
    ("triangle", "Triangle"),
    ("text", "Text"),
];

pub const DEFAULT_STYLE: &str = "chevron";

pub fn normalize_style(id: &str) -> String {
    if STYLES.iter().any(|(known, _)| *known == id) {
        id.to_string()
    } else {
        DEFAULT_STYLE.to_string()
    }
}

/// Warnings look the same in every style; only rise and fall change.
pub fn glyph(mark: char, style: &str) -> char {
    match mark {
        MARK_WARN => '⚠',
        MARK_RISE => match style {
            "arrow" => '↑',
            "triangle" => '▲',
            "chevron" => '⌃',
            _ => '▴',
        },
        MARK_FALL => match style {
            "arrow" => '↓',
            "triangle" => '▼',
            "chevron" => '⌄',
            _ => '▾',
        },
        _ => ' ',
    }
}

/// The fixed pair used everywhere the style does not apply. 1.x drew icons in
/// the menu bar only; its menu items, tooltips and clipboard always used these.
pub fn text_glyph(mark: char) -> char {
    glyph(mark, "text")
}

/// The menu bar title, which is where the chosen style shows up.
pub fn to_text_styled(value: &str, style: &str) -> String {
    value
        .chars()
        .map(|c| if is_mark(c) { glyph(c, style) } else { c })
        .collect()
}

/// Everywhere else: menu items, tooltips, the log, the clipboard and the
/// widget snapshot, none of which follow the style.
pub fn to_text(value: &str) -> String {
    to_text_styled(value, "text")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn each_style_has_its_own_pair() {
        let pairs: Vec<(char, char)> = STYLES
            .iter()
            .map(|(id, _)| (glyph(MARK_RISE, id), glyph(MARK_FALL, id)))
            .collect();
        assert_eq!(pairs.len(), 4);
        for (rise, fall) in &pairs {
            assert_ne!(rise, fall);
        }
        // Chevron and arrow must not collapse into the same rendering.
        assert_ne!(pairs[0], pairs[1]);
    }

    #[test]
    fn warnings_ignore_the_style() {
        for (id, _) in STYLES {
            assert_eq!(glyph(MARK_WARN, id), '⚠');
        }
    }

    #[test]
    fn only_the_tray_title_follows_the_style() {
        let raw = format!("{MARK_RISE}12.30");
        assert_eq!(to_text_styled(&raw, "arrow"), "↑12.30");
        // Menus, tooltips and the clipboard keep the fixed pair.
        assert_eq!(to_text(&raw), "▴12.30");
    }

    #[test]
    fn unknown_styles_fall_back() {
        assert_eq!(normalize_style("nope"), DEFAULT_STYLE);
        assert_eq!(normalize_style("arrow"), "arrow");
    }
}
